File: grind75/Dec22/207.py
from collections import defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # dfs
        # recursive
        edges = defaultdict(list)
        for course1, course2 in prerequisites:
            edges[course1].append(course2)
        
        visited = [-1] * numCourses

        def hasCycle(course):
            if visited[course] == 1:
                return False
            if visited[course] == 0:
                return True
            
            visited[course] = 0
            for edge in edges[course]:
                if hasCycle(edge):
                    return True
            visited[course] = 1
            return False
        
        for course in range(numCourses):
            if hasCycle(course):
                return False
        
        return True

File: grind75/Dec22/207.go

func canFinish(numCourses int, prerequisites [][]int) bool {
    g := make([][]int, numCourses)
    in := make([]int, numCourses)
    for _, c := range prerequisites {
        g[c[1]] = append(g[c[1]], c[0])
        in[c[0]] += 1
    }
    q := []int{}
    ret := []int{}
    for i := 0; i < numCourses; i++ {
        if in[i] == 0 {
            q = append(q, i)
            ret = append(ret, i)
        }
    }

    for len(q) > 0 {
        cq := []int{}
        lq := len(q)

        for lq > 0 {
            node := q[0]
            q = q[1:]
            lq--

            for _, next := range g[node] {
                in[next] -= 1
                if in[next] == 0 {
                    ret = append(ret, next)
                    cq = append(cq, next)
                }
            }
        }
        q = cq
    }
    return len(ret) == numCourses
}
